### ISSUE TITLE
in_dummy: increasing rate stops working after 55-60 messages/sec

### ISSUE THREAD

**Issue by cazorla19** [fluent/fluent-bit#8558]
## Bug Report

I'm trying to load-test the log aggregation system to know its limitations. Here is how my system looks like:

````
1 fluentbit-client --> 1 fluentbit-logaggregator-datacenter1 --> 1 fluentbit-logaggregator-datacenter2 --> s3-bucket
````

The client transfers logs to logag-dc1 by using `forward` output plugin, same works from logag-dc1 to logag-dc2. logag-dc2 uses `s3` output plugin writing to S3.

For load testing, I chose the `dummy` input plugin, which should be able to generate as many log messages as you want using `memory` as a `storage.type`. I tried to increase the rate gradually, multiplying the new one by 10 (1, 10, 100, 1000 etc.) and check how the system gets impacted by each next iteration. I looked at:

- log archive size in S3 bucket (using the default `upload_timeout: 10m`)
- fluent-bit process CPU and RAM usage on each stage of the system (client, logag-dc1, logag-dc2) + Node network usage
- fluent-bit logs usage on each stage of the system

I expected the log archive size increasing proportionally to the dummy rate increase, as it's the only input plugin used by the client. It worked so after increasing the rate from 1 to 10, but didn't work after increasing from 10 to 100. By setting different rates, I've found out that the log archive size stopped increasing after 55 messages/sec, which means that with every next increase, whether it's 100, 1000, or even a million, the final amount of logs stays the same.

I tried to find any bottleneck in this configuration.

- Network, CPU and Memory usage stays low and didn't even increase after changing the rate from 1 to 55, so the node capacity shouldn't be the issue. Also, the general CPU/RAM/Network node usage is far from its limits.
- There is no `Rate` limit, neither in [docs](https://docs.fluentbit.io/manual/pipeline/inputs/dummy), nor in [source code](https://github.com/fluent/fluent-bit/blob/master/plugins/in_dummy/in_dummy.c).
- There is no `Mem_Buf_Limit` for the `memory` storage type, unless you set it explicitly ([proof](https://github.com/fluent/fluent-bit/issues/669#issuecomment-630999803)), so the input plugin itself shouldn't be limited in any way
- There are no error/warn messages in fluent-bit logs on each stage of the system

Last but not least, the fluentbit-client and fluentbit-logag-dc1 are running on _Windows Server_, which might expose OS limits I'm not fully aware of.

Surely the `in_dummy` plugin must be able to generate more messages, as the limit I've reached looks ridiculous per se, so I'm seeking here for help finding an actual bottleneck.

**To Reproduce**

````
[INPUT]
    name dummy
    Tag  local.system
    Rate  100 # gradually increase from 1 to 50 then to 100 and further
    Dummy  {"message":"hello"}
 
[FILTER]
    name modify
    match  local.system
    add  host ${HOSTNAME}
 
[OUTPUT]
    name forward
    Match  *
    Host  logag-dc1
    Port  24224
````

```
[2024/03/07 17:12:25] [ info] [input:dummy:dummy.0] initializing
[2024/03/07 17:12:25] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
```

**Expected behavior**

The final log archive size increasing proportionally to the dummy rate increase.

**Your Environment**

* Version used: 2.1.10.0
* Operating System and version: Windows Server 2019

**Comment by patrick-stephens** [comment]
I don't believe the dummy plugin was ever intended to be used for perf testing but I think there are some extra config options you can use to improve things.

@pwhelan may know but you can do a `-i dummy --help` to see the config options available.

**Comment by pwhelan** [comment]
It is much better to use the `copies` parameter to attempt to test performance since that will create multiple records in a single buffer instead of attempting to submit each record individually. Both parameters can be used together as well to try and create more load.

**Comment by nokute78** [comment]
@cazorla19 We can use `counter` and `flowcounter` output plugins to count up records.
https://docs.fluentbit.io/manual/pipeline/outputs/flowcounter

Following configuration is an example to count up records per second.
Some filter/output plugin may affect throughput. Could you try to check in_dummy is bottleneck or not ?
```
[INPUT]
    name dummy
    Tag  local.system
    Rate  200 # gradually increase from 1 to 50 then to 100 and further
    Dummy  {"message":"hello"}

[OUTPUT]
    name flowcounter
    Match  *
    unit second
```

**Comment by cazorla19** [comment]
Thanks everyone for helping me out. With your advices, I was able to resolve the issue and use the dummy plugin for performance testing.

Indeed, the `Copies` parameter is much more useful than the `Rate` in my case, I was able to observe the correlation between the copies count increase and the general load increase. In the end, I removed the `Rate` from my config, because I don't really want to attempt to submit each record individually.

I was also able to see some logs from the `flowcounter` plugin, which also gave me the proof of the load increase. The only thing I've noticed is that it doesn't work as expected on Windows. While the Linux node worked just fine, the Windows node was launched with the `flowcounter`, and the fluent-bit even managed to run the tasks, as I saw in the shutdown logs, it just doesn't log the counter.

Before closing this issue as it was resolved, I just wanted to ask if this is the expected behavior from the maintainer's perspective.

**Comment by patrick-stephens** [comment]
I don't know is my answer, @cosmo0920 may know more on the Windows side but I would suggest raising as a separate issue with the easy reproducer you have to make it simpler to find/follow.
